### PR TITLE
Changed API  for property based testing (#89)

### PR DESCRIPTION
**Pull request by sergesku**
### Added

* `TestItems (p :: Purpose)` type represens the data set for script checking
* Property based testing example (test/Properties/Main.hs)

### Changed

* Changed API for property based testing:
  * Changed type and implementation for `scriptProperty`
* `Generator (p :: Purpose)` replaced with `Generator (a :: Type) (p :: Purpose)`,
    changed types for data constructors `GenForSpending` and `GenForMinting`
* Property based testing coverage check changed from 0.5% to 45%

### CODE REVIEW

File: tasty-plutus/test/Properties/Main.hs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+import Wallet.Emulator.Wallet (Wallet, fromWalletNumber, walletPubKeyHash)
+
+--------------------------------------------------------------------------------
+
+main :: IO ()
+main = defaultMain tests
+
+tests :: TestTree
+tests =
+  localOption (QuickCheckTests 100) $
```
Comment (adamczykm) [inline tasty-plutus/test/Properties/Main.hs:59]: `make test` ran 800 tests for "Output secret" property. Is the option passed properly?

File: tasty-plutus/test/Properties/Main.hs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    cb =
+      paysToSelf zero (pkh, int)
+        <> paysToPubKey userPKHash val
+    ex :: Outcome
+    ex = if int == int' then Pass else Fail
+
+simpleValidator :: (PubKeyHash, Integer) -> (Integer, Value) -> ScriptContext -> Bool
+simpleValidator (_, secret) (guess, value) sc =
+  secretUnlocked
+    && enoughtValue
```
Comment (adamczykm) [inline tasty-plutus/test/Properties/Main.hs:110]: s/enought/enough/g

File: tasty-plutus/test/Properties/Main.hs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+--------------------------------------------------------------------------------
+
+main :: IO ()
+main = defaultMain tests
+
+tests :: TestTree
+tests =
+  localOption (QuickCheckTests 100) $
+    withValidator "Property based testing" testSimpleValidator $ do
+      scriptProperty "Output secret" $ GenForSpending gen1 transform1
```
Comment (adamczykm) [inline tasty-plutus/test/Properties/Main.hs:61]: It is a little bit unclear for me what property this transform test.

File: tasty-plutus/src/Test/Tasty/Plutus/Script/Property.hs [hunk 4]
```diff
@@ -278,148 +256,124 @@ instance (Typeable p) => IsTest (PropertyTest p) where
...
+          env =
+            PropertyEnv
+              { envOpts = opts
+              , envScript = script
+              , envTestData = td
+              , envContextBuilder = cb
+              , envExpected = outcome
+              }
+       in checkCoverage
+            . cover 45.0 (outcome == Pass) "Unsuccessful validation"
```
Comment (adamczykm) [inline tasty-plutus/src/Test/Tasty/Plutus/Script/Property.hs:345]: Unsuccessful validation -> Successful validations

Comment (adamczykm) [reply tasty-plutus/src/Test/Tasty/Plutus/Script/Property.hs:345]: ?

**Review by adamczykm — CHANGES_REQUESTED** [review]
Overall I like the changes of the API, but we must communicate better wrt to the work on common libraries - you have duplicated my work from the previous week.

File: tasty-plutus/test/Properties/Main.hs (hunk 1 above)
Comment (sergesku) [reply tasty-plutus/test/Properties/Main.hs:59]: We use coverage in our testing. And it is little tricky.

What does it do? We have some `tolerance` and `certainty` for coverage.

Every `powerOfTwo * 100` tests it checks (`kLow` and `kHigh` - coefficients depending on the `certainty`):
if one of them is True - we can stop testing:
```
Successful coverage:
willson numOfClasified numOfSuccessed kLow >= tolerance * desiredCoverage

Unsuccessful coverage:
willson numOfClasified numOfSuccessed kHight < desiredCoverage
```
else - continue the testing

[About Willson law](https://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval#Wilson_score_interval).

As far as I understand, until a coverage decision is made, we cannot stop testing and QuickCheckTests will not work.

File: tasty-plutus/test/Properties/Main.hs (hunk 3 above)
Comment (sergesku) [reply tasty-plutus/test/Properties/Main.hs:61]: Hmm. Yes, my validator example is a little weird. So I built a new, simpler one.

File: tasty-plutus/test/Properties/Main.hs (hunk 1 above)
Comment (adamczykm) [reply tasty-plutus/test/Properties/Main.hs]: Ok, got it.

**Review by adamczykm — APPROVED** [review]

File: tasty-plutus/src/Test/Tasty/Plutus/Script/Property.hs [hunk 5]
```diff
@@ -137,77 +146,63 @@ import Text.Show.Pretty (ppDoc)
...
+ * For any 'TestItems' with Outcome equals 'Pass', the script succeeds; and
+ * For any 'TestItems' with Outcome equals 'Fail', the script fails.
 
  This will also check /coverage/: specifically, the property will fail unless
- the provided generation method produces roughly equal numbers of 'Pass' and
+ the provided way produces roughly equal numbers of 'Pass' and
  'Fail'-classified cases.
 
- @since 3.1
+ @since 5.0
```
Comment (adamczykm) [inline tasty-plutus/src/Test/Tasty/Plutus/Script/Property.hs:159]: I think you forgot to bump the version in cabal files.

**Review by adamczykm — CHANGES_REQUESTED** [review]

**Review by adamczykm — APPROVED** [review]

**Review by kozross — APPROVED** [review]
After this lands, we can merge `staging` into `master`, then remove the `staging` branch.

**Pull request merged**